src/inc/til/throttled_func.h

```diff
@@ -0,0 +1,206 @@
+                return hadValue;
+            }
+
+            template<typename F>
+            void modify_pending(F f)
+            {
+                std::unique_lock guard{ _lock };
+                if (_pendingRunArgs)
+                {
+                    std::apply(f, *_pendingRunArgs);
+                }
+            }
+
+            std::tuple<Args...> take()
+            {
+                std::unique_lock guard{ _lock };
+                auto pendingRunArgs = std::move(*_pendingRunArgs);
+                _pendingRunArgs.reset();
+                return pendingRunArgs;
+            }
+
+            explicit operator bool() const
+            {
+                std::shared_lock guard{ _lock };
+                return _pendingRunArgs.has_value();
+            }
+
+        private:
+            // std::mutex uses imperfect Critical Sections on Windows.
+            // --> std::shared_mutex uses SRW locks that are small and fast.
+            mutable std::shared_mutex _lock;
+            std::optional<std::tuple<Args...>> _pendingRunArgs;
+        };
+
+        template<>
+        class throttled_func_storage<>
+        {
+        public:
+            bool emplace()
+            {
+                return _isPending.exchange(true, std::memory_order_relaxed);
+            }
+
+            std::tuple<> take()
+            {
+                reset();
+                return {};
+            }
+
+            void reset()
+            {
+                _isPending.store(false, std::memory_order_relaxed);
+            }
+
+            explicit operator bool() const
+            {
+                return _isPending.load(std::memory_order_relaxed);
+            }
+
+        private:
+            std::atomic<bool> _isPending;
+        };
+    } // namespace details
+
+    template<bool leading, typename... Args>
+    class throttled_func
+    {
+    public:
+        using filetime_duration = std::chrono::duration<int64_t, std::ratio<1, 10000000>>;
+        using function = std::function<void(Args...)>;
+
+        // Throttles invocations to the given `func` to not occur more often than `delay`.
+        //
+        // If this is a:
+        // * throttled_func_leading: `func` will be invoked immediately and
+        //   further invocations prevented until `delay` time has passed.
+        // * throttled_func_trailing: On the first invocation a timer of `delay` time will
+        //   be started. After the timer has expired `func` will be invoked just once.
+        //
+        // After `func` was invoked the state is reset and this cycle is repeated again.
+        throttled_func(filetime_duration delay, function func) :
+            _func{ std::move(func) },
+            _timer{ _createTimer() }
+        {
+            const auto d = -delay.count();
+            if (d >= 0)
+            {
+                throw std::invalid_argument("non-positive delay specified");
+            }
+
+            memcpy(&_delay, &d, sizeof(d));
```

Comment:
This seems like an overly clever trick. Does the compiler not generate a sufficient copy when you use `_delay = d`?

Reply:
(Tricks like this seem like a micro-optimization, and micro-optimizations aren't _bad_ but they do deserve some scrutiny and discussion about their specific purpose :smile:)

Reply:
Actually `_delay` is a `FILETIME` and `d` is a `int64_t`.
I considered making `_delay` an `int64_t` as well, but this would require a `reinterpret_cast` between two unrelated types, which is technically undefined behavior and could break with any VS upgrade.
C++20 introduced `std::bit_cast` which would allow me to write `_delay = std::bit_cast<FILETIME>(-delay.count())`, but we don't have C++20 enabled yet (`std::bit_cast` is basically `memcpy`, but from the compiler and `constexpr`).